env.py

```diff
@@ -0,0 +1,50 @@
+"""
+Handles all environment variables set by the .env file
+
+Uses pydantic validation to allow for static and runtime type checks. The bot
+will not run unless all required fields are set.
+"""
+
+from pydantic import Field, field_validator
+from pydantic_settings import BaseSettings, SettingsConfigDict
+
+
+class _Env(BaseSettings):
+    """
+    THIS IS AN INTERNAL CLASS. DO NOT IMPORT THIS. INSTEAD USE THE `env`
+    INSTANCE FROM THIS SAME MODULE LIKE SO:
+
+    ```
+    from env import env
+    ```
+
+    A Pydantic settings class that handles validation of environment variables.
+    Reads in a .env file and ensures each variable fits the criteria. All
+    variables are required unless there is a default value assigned.
+
+    If an error occurs during parsing, then an ValidationError will throw and
+    will terminate the program.
+    """
+
+    dev_mode: bool = True
+    discord_token: str
+    discord_dev_token: str = Field(min_length=1)
+    dev_server_id: int = Field(gt=0)
+    states_server_id: int = Field(gt=0)
+    slash_command_guilds: list[int]
+    emoji_guilds: list[int]
+    pi_bot_wiki_username: str | None = None
+    pi_bot_wiki_password: str | None = None
+    mongo_url: str = Field(min_length=1)
+
```

Comment:
There are two environment variables in `docker-compose.yaml` for those who want to run the mongo database. Can those be added, too?

Reply:
Otherwise, pydantic's going to throw an error for having extra args in the `.env`

Reply:
The env vars defined are for the mongodb container, not for the python container. Afaik, bringing up the bot service using compose operates fine and reaches the ready status.

If in the future there are any explicit env vars declared in the compose file for the bot service, then yes, the validator will need to account for that.

Alternatively, I believe we make the validator pass for extra unused env vars, but still assert any explicitly defined fields. (e.g. AWS env creds)

Reply:
They are in the `docker-compose.yaml` file:
```yaml
      MONGO_INITDB_ROOT_USERNAME: ${MONGO_ROOT_USER}
      MONGO_INITDB_ROOT_PASSWORD: ${MONGO_ROOT_PASS}
```

The service may be launching okay for you because the root username/pass of your mongodb is probably just being set to an empty string.

Reply:
Oh, I see what the confusion is.

So, you use the .env file for all env vars (since Docker does can `.env` in its [env precedence](https://docs.docker.com/compose/environment-variables/envvars-precedence/)).

Running with the env variables inline or exporting them prior results in normal behavior:
```
MONGO_ROOT_USER=mongodb MONGO_ROOT_PASS=changeme docker compose up --build
```
```
export MONGO_ROOT_USER=mongodb
export MONGO_ROOT_PASS=changeme
docker compose up --build
```

Is that correct?

Reply:
Yup, I believe `docker-compose` pulls from an `.env` file when necessary

Reply:
Ok, made the change to just ignore any extra env variables (default `extra="forbid"`)